File: connection-backend/src/main/java/de/florianschmitt/system/util/ReflectionUtils.java
package de.florianschmitt.system.util;

import org.springframework.core.ResolvableType;

public class ReflectionUtils {
    public static <ENTITY> Class<ENTITY> classFromFirstGenericParameter(Object obj) {
        return classFromNthGenericParameter(obj, 0);
    }

    public static <ENTITY> Class<ENTITY> classFromSecondGenericParameter(Object obj) {
        return classFromNthGenericParameter(obj, 1);
    }

    public static <ENTITY> Class<ENTITY> classFromThirdGenericParameter(Object obj) {
        return classFromNthGenericParameter(obj, 2);
    }

    private static <ENTITY> Class<ENTITY> classFromNthGenericParameter(Object obj, int n) {
        @SuppressWarnings("unchecked")
        Class<ENTITY> entityClass = (Class<ENTITY>) ResolvableType.forClass(obj.getClass())//
                .getSuperType()//
                .resolveGeneric(n);
        return entityClass;
    }
}

File: src/main/kotlin/de/florianschmitt/system/util/ReflectionUtils.kt
package de.florianschmitt.system.util

import org.springframework.core.ResolvableType

object ReflectionUtils {
    fun <ENTITY> classFromFirstGenericParameter(obj: Any): Class<ENTITY> {
        return classFromNthGenericParameter(obj, 0)
    }

    fun <ENTITY> classFromSecondGenericParameter(obj: Any): Class<ENTITY> {
        return classFromNthGenericParameter(obj, 1)
    }

    private fun <ENTITY> classFromNthGenericParameter(obj: Any, n: Int): Class<ENTITY> {
        return ResolvableType.forClass(obj.javaClass)//
                .superType//
                .resolveGeneric(n) as Class<ENTITY>
    }
}
